akka-stream/src/main/scala/akka/stream/impl/fusing/MapAsyncPartitioned.scala

```diff
@@ -0,0 +1,254 @@
+  def apply[In, Out, Partition](
+      parallelism: Int,
+      perPartition: Int,
+      partitioner: In => Partition,
+      f: (In, Partition) => Future[Out]): GraphStage[FlowShape[In, Out]] =
+    if (parallelism < 1) throw new IllegalArgumentException("parallelism must be at least 1")
+    else if (perPartition < 1) throw new IllegalArgumentException("perPartition must be at least 1")
+    else if (perPartition >= parallelism) {
+      // no point controlling per-partition, so just use a MapAsync, which has somewhat less overhead
+      val fin = { (x: In) =>
+        f(x, partitioner(x))
+      }
+      MapAsync(parallelism, fin)
+    } else MapAsyncPartitioned(parallelism, perPartition, partitioner, f)
+
+  final class Holder[P, I, O](
+      var incoming: I,
+      var outgoing: Try[O],
+      val partition: P,
+      val cb: AsyncCallback[Holder[P, I, O]])
+      extends (Try[O] => Unit) {
+    private var cachedSupervisionDirective: OptionVal[Supervision.Directive] = OptionVal.None
+
+    def supervisionDirectiveFor(decider: Supervision.Decider, ex: Throwable): Supervision.Directive =
+      cachedSupervisionDirective match {
+        case OptionVal.Some(d) => d
+        case _ =>
+          val d = decider(ex)
+          cachedSupervisionDirective = OptionVal.Some(d)
+          d
+      }
+
+    def clearIncoming(): Unit = {
+      incoming = null.asInstanceOf[I]
+    }
+
+    def setOutgoing(t: Try[O]): Unit = {
+      outgoing = t
+    }
+
+    override def apply(t: Try[O]): Unit = {
+      setOutgoing(t)
+      cb.invoke(this)
+    }
+
+    override def toString: String = s"Holder($incoming, $outgoing, $partition)"
+  }
+
+  val NotYetThere = MapAsync.NotYetThere
+}
+
+/**
+ * Internal API
+ */
+@InternalApi
+private[akka] final case class MapAsyncPartitioned[In, Out, Partition](
+    parallelism: Int,
+    perPartition: Int,
+    partitioner: In => Partition,
+    f: (In, Partition) => Future[Out])
+    extends GraphStage[FlowShape[In, Out]] {
+  import MapAsyncPartitioned._
+
+  // duplicating the checks of the apply method since the constructor is public in the bytecode
+  if (parallelism < 1) throw new IllegalArgumentException("parallelism must be at least 1")
+  if (perPartition < 1) throw new IllegalArgumentException("perPartition must be at least 1")
+  if (perPartition >= parallelism)
+    throw new IllegalArgumentException("perPartition must be less than parallelism: consider mapAsync instead")
+
+  private val in = Inlet[In]("MapAsyncPartitioned.in")
+  private val out = Outlet[Out]("MapAsyncPartitioned.out")
+
+  override def initialAttributes = DefaultAttributes.mapAsync and SourceLocation.forLambda(f)
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
```

Comment:
As mentioned in the test comments, I think if we should no support supervision in this operator: 

It makes the already complex stage additionally complicated and for the use case we envision (message broker offset tracking) anything else than stop/fail stream doesn't really make sense and for non-default-use-case of resume on failure it would be possible to emulate with try-catch and future recover, restart doesn't really make sense at all.